model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/group/JpaGroupModelCriteriaBuilder.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.storage.jpa.group;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.map.common.StringKeyConvertor;
+import org.keycloak.models.map.storage.CriterionNotSupportedException;
+import org.keycloak.models.map.storage.jpa.group.entity.JpaGroupEntity;
+import org.keycloak.models.map.storage.jpa.JpaModelCriteriaBuilder;
+import org.keycloak.models.map.storage.jpa.hibernate.jsonb.JsonbType;
+import org.keycloak.storage.SearchableModelField;
+
+public class JpaGroupModelCriteriaBuilder extends JpaModelCriteriaBuilder<JpaGroupEntity, GroupModel, JpaGroupModelCriteriaBuilder> {
+
+    public JpaGroupModelCriteriaBuilder() {
+        super(JpaGroupModelCriteriaBuilder::new);
+    }
+
+    private JpaGroupModelCriteriaBuilder(BiFunction<CriteriaBuilder, Root<JpaGroupEntity>, Predicate> predicateFunc) {
+        super(JpaGroupModelCriteriaBuilder::new, predicateFunc);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public JpaGroupModelCriteriaBuilder compare(SearchableModelField<? super GroupModel> modelField, Operator op, Object... value) {
+        switch (op) {
+            case EQ:
+                if (modelField.equals(GroupModel.SearchableFields.REALM_ID) ||
+                    modelField.equals(GroupModel.SearchableFields.NAME)) {
+                    validateValue(value, modelField, op, String.class);
```

Comment:
Seeing all the validateValue() instances around here, I wonder if this should be done one the logical level. If SearchableModelField would receive a second generic type parameter for the fieldClass, it could be possible to create a type safe API where the query is constructed and it can be check at compile time. As this would affect a lot of classes, it should be a separate issue. 

```
public class SearchableModelField<M, T> {

    private final String name;
    private final Class<T> fieldClass;

    public SearchableModelField(String name, Class<T> fieldClass) {
        /* ... */
    }

    /* ... */

}
``` 

As modelField always contains the class, I wonder if there is really the need to pass on the class; it might be enough to pass on the number of expected arguments. This also would call for a separate issue. 

Please comment what you think is more appropriate here or if this has been discussed before. As this can be quite a big task, it also needs a decision when this should be done.

Reply:
Hi @hmlnarik - I understand that an implementation can limit the fields and operations, and potentially also parameter combinations although I don't have an example for the latter at hand. 

Given the situation above, I would assume that even the logical layer could validate that EQ mandates for exactly one parameter. When the compare() is called for REALM_ID and NAME, both specify String as their fieldClass. Would it be allowed on a logical level to pass an integer, or would we assume/require on a logical level to pass as the parameter only a type that matches the type of the field?

Reply:
I would like to have such an enhancement in. It would then be able to keep the source code tidier.

Yet I think there is a case that may be very tricky to handle in Java type system: While `EQ` is a straightforward example where such validation would work, problems arise with collection-like parameters, e.g. `IN`. Parameters of those operators might be collections, arrays as well as streams. How would this be handled?

As for the example for multiple values - attribute searchable fields require two parameters: parameter name and value. Since java has no varags support for type variables, I suspect the solution would be a bit hacky involving number of warnings, though I'm happy to be convinced otherwise.

I'm open for suggestions how to handle those, as long as the solution would be straightforward.

As a side note: I'm not convinced the `fieldClass` should remain amongst properties the searchable fields. I don't think it's ever used and while introduced with a good intention similar to the suggestion here, it turned out to be impractical to leverage it.

Reply:
I suppose the fieldClass could be leveraged for run time checking on the logical level where type checking is not possible. 

So it's now "challenge accepted" and I'll see what's possible, and it will be handled outside of this PR. 

A short comment on priorities would be nice; it would change the signature of SearchableModelField in lots of places, and I see the benefit that we have common checks only once in the logical layer instead in each storage; and then each store can focus on additional restrictions on top of the common checks.

Reply:
> A short comment on priorities would be nice [...]

Priority-wise I don't see this as a priority. This is certainly a nice-to-have rather than must-have for functionalit:
- During runtime, such a check would only be able to throw an exception which is the situation we're in already.
- It would be useful if it could be used for testing during compile time.

Reply:
resolved.